datalad_debian/update_reprepro_repository.py

```diff
@@ -0,0 +1,226 @@
+            ['log', '-1', '--format=%H'], files='www')
+        lgr.debug('Using archive update ref %r', last_update_hexsha)
+
+        dist_subdataset_relpath = reprepro_ds.subdatasets(
+            'distributions',
+            result_xfm='relpaths',
+            result_renderer="disabled")
+
+        updated_dists = [
+            d for d in reprepro_ds.diff(
+                'distributions',
+                fr=last_update_hexsha,
+                result_xfm='relpaths',
+                result_renderer='disabled')
+            if d in dist_subdataset_relpath
+        ]
+        # TODO could be done in parallel
+        for ud in updated_dists:
+            yield from _get_updates_from_dist(
+                reprepro_ds,
+                Path(ud),
+                last_update_hexsha,
+            )
+
+
+def _get_updates_from_dist(ds, dpath, ref):
+    # TODO option to drop distributions that were not present locally before?
+    # make sure the distribution dataset is present locally
+    lgr.debug('Updating from %s', dpath)
+    dist_ds = ds.get(
+        path=dpath,
+        get_data=False,
+        result_xfm='datasets',
+        result_renderer='disabled',
+        return_type='item-or-list',
+    )
+    updated_pkg_datasets = [
+        # we must use `ds` again to keep the validity of `ref`
+        pkg_ds for pkg_ds in ds.diff(
+            fr=ref,
+            path=dpath / 'packages',
+            recursive=True,
+            recursion_limit=1,
+            result_xfm='datasets',
+            result_renderer='disabled',
+        )
+        # we are not interested in the distribution dataset here
+        if pkg_ds != dist_ds
+    ]
+    # TODO could be done in parallel
+    for up in updated_pkg_datasets:
+        yield from _get_updates_from_pkg(
+            ds,
+            dist_ds.pathobj.name,
+            up,
+            ref,
+        )
+
+
+def _get_updates_from_pkg(ds, dist_codename, pkg_ds, ref):
+    # TODO option to drop packages that were not present locally before?
+    # make sure the package dataset is present locally
+    lgr.debug('Updating from %s', pkg_ds.pathobj.relative_to(ds.pathobj))
+    ds.get(
+        path=pkg_ds.pathobj,
+        get_data=False,
+        result_renderer='disabled',
+        return_type='item-or-list',
+    )
+    updated_files = [
+        # we must use `ds` again to keep the validity of `ref`
+        Path(r['path']) for r in ds.diff(
+            fr=ref,
+            path=pkg_ds.pathobj,
+            recursive=True,
+            recursion_limit=2,
+            result_renderer='disabled',
+        )
+        # we can handle three types of files
+        # - changes files from builds of any kind
+        # - dsc of source packages
+        # - lonely debs
+        if r.get('state') in ('added', 'modified')
+        and Path(r['path']).suffix in ('.changes', '.dsc', '.deb')
+    ]
+    if not updated_files:
+        return
+    # TODO option to give a single commit across all updates?
```

Comment:
I think high commit count will be a problem for projects on the scale of Debian (given current limitations of git).

Reply:
Yes. That is the rational behind the TODO